File: prototype-lab/projects/project-records-room-poc/static/coordination.js
/* Real receipt-backed coordination; no simulated agent presence or execution. */
(()=>{
  const labels={requested:'Requested · awaiting pickup',picked_up:'Picked up',result_submitted:'Result submitted · awaiting acknowledgment',acknowledged:'Acknowledged',cancelled:'Cancelled'};
  const panel=element('section');panel.id='coordination-panel';
  const heading=element('h4','Requests & handoffs'),add=element('button','＋ Request','text-button');add.id='new-coordination';heading.append(add);
  const list=element('div');list.id='coordination-items';panel.append(heading,element('p','Pickup and results are recorded by the assigned identity. Posting does not launch an agent.','muted'),list);
  document.querySelector('.room-brief').prepend(panel);
  const brief=element('section'),briefhead=element('h4','Executive briefings'),capture=element('button','＋ Snapshot','text-button');capture.id='new-snapshot';briefhead.append(capture);const snapshots=element('div');snapshots.id='executive-snapshots';brief.append(briefhead,element('p','Selected source records, frozen at capture time. This conversation stays separate from the working session.','muted'),snapshots);panel.after(brief);
  const inbox=element('button','Requests for you','secondary');inbox.id='coordination-inbox';inbox.classList.add('hidden');document.querySelector('.environment').after(inbox);
  let running=false,lastRoom=null,lastAuth=null,signature='',inboxItems=[],seen=new Set();
  const button=(label,fn)=>{const b=element('button',label,'text-button');b.type='button';b.onclick=fn;return b;};
  function actions(item){
    const out=[];if(state.room?.state!=='active')return out;
    if(item.assignee===state.me.id&&item.state==='requested')out.push(['pickup','Acknowledge pickup']);
    if(item.assignee===state.me.id&&item.state==='picked_up')out.push(['submit','Submit result']);
    if(item.kind==='owner_input'&&state.me.id==='robert'&&['requested','picked_up'].includes(item.state))out.push(['answer','Answer request']);
    if(item.requester===state.me.id&&item.state==='result_submitted')out.push(['acknowledge','Acknowledge result']);
    if(['requested','picked_up','result_submitted'].includes(item.state)&&[item.requester,'robert'].includes(state.me.id))out.push(['cancel','Cancel request']);
    return out;
  }
  function change(item,action,label){
    const context=roomContext();modal(label,[field('body','Response / evidence','textarea'),element('p','This records your own response. It does not execute the requested work.')],async data=>{context.check();await write(`/api/v1/rooms/${context.id}/coordination/${item.id}`,{action,body:data.body,version:item.version});signature='';await refresh();},label);
  }
  function renderItem(item){
    const node=element('article',undefined,'coordination-card');node.dataset.item=item.id;node.append(element('strong',item.title),element('small',`${item.kind} · ${labels[item.state]}`),element('p',item.body),element('small',`${item.requester} → ${item.assignee}`));
    if(item.source_snapshot)node.append(element('small',`Explicit handoff from snapshot ${item.source_snapshot}`));
    const audit=element('details');audit.append(element('summary','History & evidence'));for(const step of item.steps)audit.append(element('p',`${date(step.created)} · ${step.actor} · ${step.action}\n${step.body}`));node.append(audit);
    for(const [action,label] of actions(item))node.append(button(label,()=>change(item,action,label)));return node;
  }
  add.onclick=()=>{
    const context=roomContext();modal('Request work or input',[field('kind','Request type','select','review',[['review','Review'],['handoff','Handoff'],['owner_input','Robert input needed']]),field('title','Title'),field('body','Question, candidate reference, or handoff','textarea'),field('assignee','Assigned participant','select','robert',state.room.members.filter(m=>m.role==='contributor').map(m=>[m.id,m.label])),element('p','For Robert input, select Robert. A request is not a running agent or an approval.')],async data=>{context.check();await write(`/api/v1/rooms/${context.id}/coordination`,data);signature='';await refresh();},'Record request');
  };
  capture.onclick=async()=>{
    const context=roomContext();await refreshRooms();context.check();const choices=state.rooms.filter(r=>r.id!==context.id).map(r=>[r.id,r.title]);
    if(!choices.length){toast('Create a separate working session first.',true);return;}
    modal('Choose source working session',[field('source','Source session','select',choices[0][0],choices)],async data=>{
      context.check();const source=await api(`/api/v1/rooms/${data.source}`);context.check();
      setTimeout(()=>{try{context.check();}catch{return;}const checks=element('div');checks.className='snapshot-selection';
        for(const event of source.events.slice(-100)){const label=element('label'),check=element('input');check.type='checkbox';check.name='source-record';check.value=event.id;label.append(check,document.createTextNode(`${event.actor_label} · ${date(event.created)} · ${(event.presentation?.text||event.body).slice(0,200)}`));checks.append(label);}
        modal('Select records to disclose',[element('p',`Copy selected records from “${source.title}” into this session for its participants to read. The working session remains unchanged. Only the latest 100 source records are offered here.`),checks,field('disclosure','I authorize sharing these selected records with this session','checkbox',false)],async()=>{context.check();const ids=[...checks.querySelectorAll('input:checked')].map(n=>n.value);if(!$('field-disclosure').checked)throw new Error('Acknowledge the disclosure.');await write(`/api/v1/rooms/${context.id}/executive-snapshots`,{source:source.id,event_ids:ids,disclosure_acknowledged:true});signature='';await refresh();},'Capture snapshot');},0);
    },'Choose records');
  };
  function handoff(snapshot){
    const context=roomContext();api(`/api/v1/rooms/${snapshot.source}`).then(source=>{
      context.check();modal('Send explicit handoff to working session',[field('title','Handoff title'),field('body','Authorized next steps','textarea'),field('assignee','Assigned participant','select','robert',source.members.filter(m=>m.role==='contributor').map(m=>[m.id,m.label])),element('p',`This records a linked handoff in “${source.title}”. The assignee must acknowledge pickup; no runtime is launched.`)],async data=>{context.check();await write(`/api/v1/rooms/${snapshot.source}/coordination`,{...data,kind:'handoff',source_snapshot:snapshot.id});toast('Handoff recorded in working session; awaiting assignee pickup.');},'Send handoff');
    }).catch(error=>toast(error.message,true));
  }
  async function refresh(){
    if(running||!state.me||state.view!=='rooms'||!state.room)return;
    const room=state.room.id,epoch=state.authEpoch,generation=state.navigation;running=true;
    const current=()=>state.me&&state.authEpoch===epoch&&state.room?.id===room&&state.navigation===generation&&state.view==='rooms';
    if(lastRoom!==room||lastAuth!==epoch){clear(list);clear(snapshots);signature='';lastRoom=room;lastAuth=epoch;}
    add.disabled=state.room.state!=='active'||!state.room.members.some(m=>m.id===state.me.id&&m.role==='contributor');capture.classList.toggle('hidden',state.me.id!=='robert');capture.disabled=state.room.state!=='active';
    try{const data=await api(`/api/v1/rooms/${room}/coordination`);if(!current())return;const next=JSON.stringify([data,state.room.state]);if(next===signature)return;signature=next;clear(list);clear(snapshots);
      for(const item of data.items)list.append(renderItem(item));if(!data.items.length)list.append(element('p','No recorded requests.','muted'));
      for(const snap of data.snapshots){const node=element('article',undefined,'coordination-card');node.append(element('strong',`Snapshot · ${date(snap.created)}`),element('small',`${snap.records.length} selected records · through source sequence ${snap.through_seq}`));
        const details=element('details');details.append(element('summary','Read briefing records'));for(const record of snap.records){details.append(element('p',`${record.actor} · ${date(record.created)}\n${record.text}`));if(record.warning)details.append(element('p',record.warning,'agent-warning'));}node.append(details);
        if(state.me.id==='robert'){node.append(button('Open source session',()=>{location.hash=`room/${snap.source}`;}),button('Send handoff back',()=>handoff(snap)));}snapshots.append(node);
      }
    }catch(error){if(current()){clear(list);clear(snapshots);list.append(element('p','Coordination unavailable. Retrying…','muted'));signature='';}}finally{running=false;}
  }
  inbox.onclick=()=>{modal('Requests for your attention',inboxItems.map(item=>{const card=element('div',undefined,'coordination-card');card.append(element('strong',item.title),element('p',`${labels[item.state]} · ${item.body}`),button('Join session',()=>{$('modal').close();location.hash=`room/${item.room}`;}));return card;}),async()=>{},'Close');};
  async function refreshInbox(){
    if(!state.me){inbox.classList.add('hidden');inboxItems=[];seen.clear();clear(list);clear(snapshots);signature='';lastAuth=null;return;}
    const epoch=state.authEpoch;
    try{const data=await api('/api/v1/coordination/inbox');if(epoch!==state.authEpoch||!state.me)return;inboxItems=data.items;inbox.classList.toggle('hidden',!data.items.length);inbox.textContent=`Requests for you (${data.items.length})`;
      const fresh=data.items.filter(i=>!seen.has(`${i.id}:${i.version}`));for(const item of data.items)seen.add(`${item.id}:${item.version}`);if(fresh.length)toast(`${fresh.length} request(s) need your attention. Open “Requests for you”.`);
    }catch{if(epoch===state.authEpoch){inboxItems=[];inbox.classList.add('hidden');}}
  }
  window.addEventListener('records-signout',()=>{clear(list);clear(snapshots);inbox.classList.add('hidden');inboxItems=[];seen.clear();signature='';lastAuth=null;});
  setInterval(()=>{if(!state.me){clear(list);clear(snapshots);inbox.classList.add('hidden');inboxItems=[];seen.clear();}else refresh();},2500);
  setInterval(refreshInbox,10000);window.addEventListener('hashchange',()=>{refresh();refreshInbox();});refresh();refreshInbox();
})();

File: prototype-lab/projects/project-records-room-poc/coordination.py
"""Receipt-backed collaboration. Records requests; never launches a runtime."""
import json
from platform_access import AccessError
from store import Problem, canonical, now, string, uid


class Coordination:
    def __init__(self, store):
        self.store=store
        with store.connect() as db:
            db.executescript('''
                CREATE TABLE IF NOT EXISTS coordination_items(
                    id TEXT PRIMARY KEY,room TEXT NOT NULL,kind TEXT NOT NULL,
                    title TEXT NOT NULL,body TEXT NOT NULL,requester TEXT NOT NULL,
                    assignee TEXT NOT NULL,state TEXT NOT NULL,version INTEGER NOT NULL,
                    created TEXT NOT NULL,updated TEXT NOT NULL,source_snapshot TEXT);
                CREATE TABLE IF NOT EXISTS coordination_steps(
                    id TEXT PRIMARY KEY,item TEXT NOT NULL,actor TEXT NOT NULL,
                    action TEXT NOT NULL,body TEXT NOT NULL,created TEXT NOT NULL);
                CREATE TABLE IF NOT EXISTS executive_snapshots(
                    id TEXT PRIMARY KEY,room TEXT NOT NULL,source TEXT NOT NULL,
                    actor TEXT NOT NULL,created TEXT NOT NULL,through_seq INTEGER NOT NULL,
                    records TEXT NOT NULL);
            ''')
            if 'source_snapshot' not in {r['name'] for r in db.execute('PRAGMA table_info(coordination_items)')}:
                db.execute('ALTER TABLE coordination_items ADD COLUMN source_snapshot TEXT')

    def _active(self, db, actor, room):
        current=self.store.access(db,actor,room,True)
        if current['state']!='active':raise Problem('Session is not recording',409)

    def create(self, actor, key, room, payload):
        if set(payload)-{'kind','title','body','assignee','source_snapshot'} or not {'kind','title','body','assignee'}<=set(payload):raise Problem('Provide kind, title, body and assignee')
        kind=payload['kind']
        if kind not in {'review','handoff','owner_input'}:raise Problem('Unknown coordination kind')
        title=string(payload['title'],'Title',160);body=string(payload['body'],'Request',16000)
        assignee=string(payload['assignee'],'Assignee',60)
        if assignee==actor:raise Problem('Choose a different participant; requests require an independent recipient')
        if assignee=='cos-dev':raise Problem('Chief of Staff replies to messages; it cannot pick up coordination requests yet')
        if kind=='owner_input' and assignee!='robert':raise Problem('Owner input is addressed to Robert')
        def action(db):
            self._active(db,actor,room)
            if not db.execute("SELECT 1 FROM members WHERE room=? AND actor=? AND role='contributor'",(room,assignee)).fetchone():
                raise Problem('Assignee needs contributor membership in this session',403)
            snapshot=payload.get('source_snapshot')
            if snapshot is not None:
                snapshot=string(snapshot,'Source snapshot',100)
                self.store.owner(actor)
                link=db.execute('SELECT * FROM executive_snapshots WHERE id=? AND source=?',(snapshot,room)).fetchone()
                if not link or kind!='handoff':raise Problem('Handoff must reference an executive snapshot of this working session',403)
                self.store.access(db,actor,link['room'])
            item=uid();at=now()
            db.execute('INSERT INTO coordination_items VALUES(?,?,?,?,?,?,?,?,?,?,?,?)',
                       (item,room,kind,title,body,actor,assignee,'requested',1,at,at,snapshot))
            db.execute('INSERT INTO coordination_steps VALUES(?,?,?,?,?,?)',(uid(),item,actor,'requested',body,at))
            self.store._event(db,room,actor,'coordination',f'{kind}: {title} · requested from {assignee}',
                              origin={'source_application':'minimoi_coordination','mode':'recorded_request'})
            return self._item(db,item)
        return self.store.mutate(actor,key,dict(op='coordination_create',room=room,payload=payload),action,room)

    def _item(self, db, item):
        row=db.execute('SELECT * FROM coordination_items WHERE id=?',(item,)).fetchone()
        if not row:raise Problem('Request not found',404)
        data=dict(row)
        data['steps']=[dict(r) for r in db.execute('SELECT * FROM coordination_steps WHERE item=? ORDER BY created,rowid',(item,))]
        data['execution']='Recorded workflow only; no runtime launched'
        return data

    def transition(self, actor, key, room, item, payload):
        if set(payload)!={'action','body','version'}:raise Problem('Provide action, body and version')
        verb=payload['action'];body=string(payload['body'],'Evidence / response',16000)
        if type(payload['version']) is not int:raise Problem('Version must be an integer')
        def action(db):
            self._active(db,actor,room);row=self._item(db,item)
            if row['room']!=room:raise Problem('Request not found',404)
            if row['version']!=payload['version']:raise Problem('Request changed; refresh before acting',409)
            # Assigned identity must speak for itself; owner cannot fake a pickup.
            if verb=='pickup' and row['state']=='requested' and actor==row['assignee']:state='picked_up'
            elif verb=='submit' and row['state']=='picked_up' and actor==row['assignee']:state='result_submitted'
            elif verb=='answer' and row['kind']=='owner_input' and row['state'] in {'requested','picked_up'} and actor=='robert':state='result_submitted'
            elif verb=='acknowledge' and row['state']=='result_submitted' and actor==row['requester'] and actor!=row['assignee']:state='acknowledged'
            elif verb=='cancel' and row['state'] not in {'acknowledged','cancelled'} and actor in {row['requester'],'robert'}:state='cancelled'
            else:raise Problem('This identity cannot perform that transition from the current state',403)
            at=now();db.execute('UPDATE coordination_items SET state=?,version=version+1,updated=? WHERE id=?',(state,at,item))
            db.execute('INSERT INTO coordination_steps VALUES(?,?,?,?,?,?)',(uid(),item,actor,verb,body,at))
            self.store._event(db,room,actor,'coordination',f"{row['title']} · {state}",
                              origin={'source_application':'minimoi_coordination','mode':'recorded_transition'})
            return self._item(db,item)
        return self.store.mutate(actor,key,dict(op='coordination_transition',room=room,item=item,payload=payload),action,room)

    def list(self, actor, room):
        with self.store.connect() as db:
            db.execute('BEGIN');self.store.access(db,actor,room)
            return dict(items=[self._item(db,r['id']) for r in db.execute('SELECT id FROM coordination_items WHERE room=? ORDER BY updated DESC',(room,))],
                        snapshots=[dict(r,records=json.loads(r['records'])) for r in db.execute('SELECT * FROM executive_snapshots WHERE room=? ORDER BY created DESC',(room,))])

    def inbox(self, actor):
        items=[]
        with self.store.connect() as db:
            db.execute('BEGIN')
            rows=db.execute("SELECT * FROM coordination_items WHERE (assignee=? AND state IN ('requested','picked_up')) OR (requester=? AND state='result_submitted') ORDER BY updated DESC",(actor,actor)).fetchall()
            for row in rows:
                try:self.store.access(db,actor,row['room'])
                except (Problem,AccessError) as error:
                    if error.status in {403,404}:continue
                    raise
                # Installation grants are checked by access as well as membership.
                items.append(self._item(db,row['id']))
                if len(items)>=100:break
        return dict(items=items,limit=100)

    def snapshot(self, actor, key, room, payload):
        self.store.owner(actor)
        if set(payload)!={'source','event_ids','disclosure_acknowledged'} or payload['disclosure_acknowledged'] is not True:
            raise Problem('Explicitly acknowledge disclosure of selected source records to this session')
        source=string(payload['source'],'Source session',100);ids=payload['event_ids']
        if source==room:raise Problem('Choose a separate executive session')
        if not isinstance(ids,list) or not 1<=len(ids)<=100 or any(not isinstance(i,str) for i in ids) or len(set(ids))!=len(ids):
            raise Problem('Select 1–100 distinct source records')
        def action(db):
            self._active(db,actor,room);self.store.access(db,actor,source)
            records=[]
            for eid in ids:
                row=db.execute('SELECT e.*,p.kind AS actor_kind FROM events e JOIN principals p ON p.id=e.actor WHERE e.room=? AND e.id=?',(source,eid)).fetchone()
                if not row:raise Problem('Selected source record unavailable',404)
                from contribution_view import connector_view
                event=dict(row);event['origin']=json.loads(event['origin']) if event['origin'] else None
                view=connector_view(db,event)
                records.append(dict(id=row['id'],seq=row['seq'],actor=row['actor'],kind=row['kind'],created=row['created'],
                    text=view['text'] if view else row['body'],warning=view['warning'] if view else None))
            records.sort(key=lambda r:r['seq']);sid=uid();at=now()
            db.execute('INSERT INTO executive_snapshots VALUES(?,?,?,?,?,?,?)',(sid,room,source,actor,at,records[-1]['seq'],canonical(records)))
            self.store._event(db,room,actor,'executive_snapshot',f'Executive briefing snapshot · {len(records)} selected records · captured {at}',
                              origin={'source_application':'minimoi_executive_snapshot','mode':'explicit_disclosure'})
            return dict(snapshot_id=sid,source=source,destination=room,created=at,through_seq=records[-1]['seq'],records=records,
                        coverage='Owner-selected records only; not a full-session summary')
        return self.store.mutate(actor,key,dict(op='executive_snapshot',room=room,payload=payload),action,room)


def export_briefing(db, room, event):
    """Expand the immutable selected records in exports without rewriting history."""
    body=event['body']
    if event['kind']!='executive_snapshot':return body
    if not db.execute("SELECT 1 FROM sqlite_master WHERE type='table' AND name='executive_snapshots'").fetchone():return body
    for snapshot in db.execute('SELECT * FROM executive_snapshots WHERE room=?',(room,)):
        # The original marker contains the exact snapshot timestamp, including microseconds.
        marker=f"Executive briefing snapshot · {len(json.loads(snapshot['records']))} selected records · captured {snapshot['created']}"
        if body!=marker:continue
        lines=[body,'',f"Source session: {snapshot['source']}",f"Snapshot: {snapshot['id']}",
               'Coverage: owner-selected records only; quoted speakers are not the submitting identity.']
        for record in json.loads(snapshot['records']):
            lines.extend(['',f"Quoted speaker: {record['actor']} · source record {record['id']} · {record['created']}",record['text']])
            if record.get('warning'):lines.append(record['warning'])
        return '\n'.join(lines)
    return body
